Add utility function to save delayed writer results in one go

pnuu opened the pull request:
This PR adds an slightly modified utility function from Polar2Grid that computes/saves all given delayed results from (multiple) calls to `res = scn.save_datasets(..., compute=False)` in one go.

 - [x] Tests added
 - [x] Tests passed
 - [x] Passes ``git diff origin/master **/*py | flake8 --diff``
 - [x] Fully documented

djhoese commented on satpy/writers/__init__.py:
```diff
@@ -381,6 +381,46 @@ def to_image(dataset, copy=False, **kwargs):
+                continue
+            for i in range(num):
+                sources.append(res[0][i])
+                targets.append(res[1][i])
```
I'm not sure it matters performance-wise but couldn't this be replaced with `extend` and `zip`? Then you could probably also remove the check for zero length.

djhoese commented on doc/source/writers.rst:
```diff
@@ -101,4 +101,20 @@ The above example can be used in enhancements YAML config like this:
+================================
+
+As mentioned earlier, it is possible to save ``Scene`` datasets directly
+using :meth:`~satpy.scene.Scene.save_datasets` method:.  However,
```
Extra colon after method?

pnuu commented:
Added tests for using `compute_writer_results()` for `simple_image` and `geotiff` writer output. Also simplified the source/target collection as suggested. For `mitiff` and `cf` writers there are commented-out placeholders, as I didn't figure out what metadata they need.

djhoese commented on satpy/tests/test_writers.py:
```diff
@@ -253,13 +256,15 @@ def test_enhance_with_sensor_entry2(self):
 
 
 class TestYAMLFiles(unittest.TestCase):
+
```
I'm 99% sure this is an undesired non-PEP8 change (blank line between class declaration and class docstring).

pnuu replied:
Hmm. Must be something `autopep8` did, as I didn't touch the existing tests. Removed the blank lines.

pnuu commented:
Hmm. Maybe. Yes. Then `save_datasets()` could call `res = self.save_dataset(composite, ..., compute=False)`, collect the results and then call `compute_writer_results(results)`.

My use case for this utility function is in operations (trollflow and trollflow-sat) for saving several composites -- possibly from several different resampled scenes -- that all might have different filename patterns and writers.

pnuu commented:
I adjusted `satpy.writers.Writer.save_datasets()` so that it also uses this utility function. I needed to adjust one unit test for `simple_image` writer as now `res = scn.save_datasets(..., compute=False)` always returns a `list` of one or more objects of either `dask.delayed` or `tuple` of two lists (sources and targets), so a direct `res.compute()` won't work.

I didn't find anything in the writer documentation mentioning using `res.compute()`, so I think this won't affect anyone. Maybe need to check the PyTroll examples, too...

pnuu commented:
The change should be safe also for `pytroll-examples`. And the documentation addition I already did uses the `compute_writer_results()` utility function, so it should be the most obvious for the users. I hope.

djhoese commented on satpy/tests/test_writers.py:
```diff
@@ -30,7 +30,7 @@
 import xarray as xr
 
-from satpy.writers import show, to_image
+from satpy.writers import show, to_image, compute_writer_results
```
Could you move the satpy imports to be inside the tests so that import errors/failures on satpy's side of things don't cause all tests to not run?
